rpc: panic in per-peer metrics

Hm, it looks like we hit some other panics here too, when adding RPC metrics around here:

https://github.com/cockroachdb/cockroach/blob/caaa7d8c2a75012589e05cc5db582a796d0ea2c1/pkg/rpc/peer.go#L178

@tbg made some recent changes here in #99191.

```
panic: child [3 10.142.1.101:26257 default] already exists [recovered]
        panic: child [3 10.142.1.101:26257 default] already exists

goroutine 1399700 [running]:
panic({0x5a2a400, 0xc017b2f7a0})
        GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc00d05b388 sp=0xc00d05b2c8 pc=0x49e43a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xe9a0aa?, {0x766e250, 0xc0146360f0})
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a fp=0xc00d05b3d0 sp=0xc00d05b388 pc=0x11cba4a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2.3()
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x2e fp=0xc00d05b3f8 sp=0xc00d05b3d0 pc=0x11cd00e
runtime.deferCallSave(0xc00d05b4c8, 0xc018dd5f90?)
        GOROOT/src/runtime/panic.go:796 +0x88 fp=0xc00d05b408 sp=0xc00d05b3f8 pc=0x49e028
runtime.runOpenDeferFrame(0xc00d93e1e0?, 0xc0009be140)
        GOROOT/src/runtime/panic.go:769 +0x1a5 fp=0xc00d05b450 sp=0xc00d05b408 pc=0x49de45
panic({0x5a2a400, 0xc017b2f7a0})
        GOROOT/src/runtime/panic.go:884 +0x212 fp=0xc00d05b510 sp=0xc00d05b450 pc=0x49e292
github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric.(*childSet).add(0xc0098cecc0, {0x7663540, 0xc009adf860})
        github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric/agg_metric.go:107 +0x1d7 fp=0xc00d05b5f8 sp=0xc00d05b510 pc=0x1a9da17
github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric.(*AggGauge).AddChild(...)
        github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric/gauge.go:87
github.com/cockroachdb/cockroach/pkg/rpc.(*Metrics).acquire(0xc0005b7d38, {{0xc000a7dec0?, 0x56869e0?}, 0x10e58f0?, 0xc0?})
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/metrics.go:189 +0x31f fp=0xc00d05b6e0 sp=0xc00d05b5f8 pc=0x1b0715f
github.com/cockroachdb/cockroach/pkg/rpc.(*Context).newPeer(0xc0005b7c00, {{0xc000a7dec0?, 0xc000a7dec0?}, 0xb7?, 0x0?})
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/peer.go:178 +0x65 fp=0xc00d05b888 sp=0xc00d05b6e0 pc=0x1b079e5
github.com/cockroachdb/cockroach/pkg/rpc.(*Context).grpcDialNodeInternal(0xc0005b7c00, {0xc000a7dec0, 0x12}, 0x3, 0x0)
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:2093 +0x453 fp=0xc00d05bb08 sp=0xc00d05b888 pc=0x1b03dd3
github.com/cockroachdb/cockroach/pkg/rpc.(*Context).GRPCDialNode(0xc0122f4c10?, {0xc000a7dec0?, 0xc0008e68c0?}, 0x3?, 0x0?)
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:2050 +0x12e fp=0xc00d05bb90 sp=0xc00d05bb08 pc=0x1b038ce
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial(0xc0038fb860, {0x766e250, 0xc009adf7a0}, 0x30?, {0x7641a50?, 0xc004902698}, 0x1, 0x0?)
        github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:170 +0xb1 fp=0xc00d05bc40 sp=0xc00d05bb90 pc=0x1b1afd1
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).Dial(0xc0038fb860, {0x766e250, 0xc009adf7a0}, 0x5c4aba0?, 0x0?)
        github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:103 +0x11c fp=0xc00d05bcb8 sp=0xc00d05bc40 pc=0x1b1aa5c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).startProcessNewQueue.func2({0x766e250, 0xc009adf7a0})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:925 +0x226 fp=0xc00d05be18 sp=0xc00d05bcb8 pc=0x3994426
runtime/pprof.Do({0x766e250?, 0xc0146360f0?}, {{0xc016fd0fe0?, 0xc0122f4eb8?, 0xc63326?}}, 0xc01a845960)
        GOROOT/src/runtime/pprof/runtime.go:40 +0xa3 fp=0xc00d05be88 sp=0xc00d05be18 pc=0xe9a003
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).startProcessNewQueue.func3({0x766e250, 0xc0146360f0})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:947 +0x1c8 fp=0xc00d05bf30 sp=0xc00d05be88 pc=0x39941a8
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146 fp=0xc00d05bfe0 sp=0xc00d05bf30 pc=0x11ccec6
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc00d05bfe8 sp=0xc00d05bfe0 pc=0x4d35c1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
```

_Originally posted by @erikgrinaker in https://github.com/cockroachdb/cockroach/issues/105260#issuecomment-1602402172_

--- a/pkg/rpc/peer.go
+++ b/pkg/rpc/peer.go
@@ -846,26 +846,21 @@ func (p *peer) maybeDelete(ctx context.Context, now time.Time) {
 
 	log.VEventf(ctx, 1, "deleting peer")
 
-	// Lock order: map, then peer. But here we can do better and
-	// not hold both mutexes at the same time.
+	// Lock order: map, then peer. We need to lock both because we want
+	// to atomically release the metrics while removing from the map[1][2].
 	//
-	// Release metrics in the same critical section as p.deleted=true
-	// to make sure the metrics are not updated after release, since that
-	// causes the aggregate metrics to drift.
-	//
-	// We delete from the map first, then mark the peer as deleted. The converse
-	// works too, but it makes for flakier tests because it's possible to see the
-	// metrics change but the peer still being in the map.
-
+	// [1]: see https://github.com/cockroachdb/cockroach/issues/105335
+	// [2]: Releasing in one critical section with p.deleted=true ensures
+	//      that the metrics are not updated after release, which would
+	//      otherwise cause the aggregate metrics to drift away from zero
+	//      permanently.
 	p.peers.mu.Lock()
+	defer p.peers.mu.Unlock()
 	delete(p.peers.mu.m, p.k)
-	p.peers.mu.Unlock()
-
 	p.mu.Lock()
+	defer p.mu.Unlock()
 	p.mu.deleted = true
 	p.peerMetrics.release()
-	p.mu.Unlock()
-
 }
 
 func launchConnStateWatcher(

